### ISSUE TITLE
ERROR ITMS-90208 on delivery to App Store

### ISSUE THREAD

**Issue by minosvasilias** [alessandrofama/wwise-godot-integration#55]
Attempting to deliver an iOS archive to the App Store will fail with the following error, thrown on upload via Xcode or Transporter.

```
ERROR ITMS-90208: "Invalid Bundle. The bundle APP_NAME.app/Frameworks/libWwiseGDNative.framework does not support the minimum OS Version specified in the Info.plist."
```

I am using the latest `1.5.0` release.
The default `MinimumOSVersion` value specified in `libWwiseGDNative.framework/Info.plist` is `10.0`, however changing this to different values (tested: `9.0`, `10.0`, `11.0`, `15.0`) makes no difference. My project generally requires a deployment target of `11.0` due to another framework dependency, but i did remove that one to test purely the impact of the Wwise framework. Both the Project-scope and target-scope deployment target values of the project matched the `libWwiseGDNative.framework` value on each test.

Most topics i can find regarding this error relate to Flutter apps, with the solution being to increase the `MinimumOSVersion` in the framework `Info.plist`. But that seems to not do the trick here.

Any ideas what could be done to resolve this? I am intending to integrate Wwise into an existing production app, and currently this issue is preventing me from doing so. Help would be greatly appreciated. Thank you. :)

**Comment by minosvasilias** [comment]
I seem to have been able to resolve this for now. The steps followed were the following:

- Edit `addons/wwise/bin/wwise-gdnative.gdnlib` to ensure release libraries are exported: 
```diff

-iOS.armv7="res://addons/wwise/bin/iOS/debug/armv7/libWwiseGDNative.dylib"
-iOS.arm64="res://addons/wwise/bin/iOS/debug/arm64/libWwiseGDNative.dylib"
+iOS.armv7="res://addons/wwise/bin/iOS/release/armv7/libWwiseGDNative.dylib"
+iOS.arm64="res://addons/wwise/bin/iOS/release/arm64/libWwiseGDNative.dylib"
```
- Export project
- Edit `dylibs/addons/wwise/bin/iOS/release/arm64/libWwiseGDNative.framework/Info.plist` in exported Xcode project
```diff
<key>MinimumOSVersion</key>
-<string>10.0</string>
+<string>15.4</string>
```
- Archive and test upload to App Store

`15.4` is the absolute latest target OS version you could set at time of writing, and i do not wish to actually target it. As described in the OP however, even `15.0` appears to be too low. That being said, uploading to the App Store did not throw the usual warning about increasing the min OS version, and Testflight also displays the correct version specified for the project: 


<img width="275" alt="Screenshot 2022-06-05 at 19 37 37" src="https://user-images.githubusercontent.com/44006014/172063168-6121463c-6179-4dd9-8d81-0d9686b13037.png">


So i'd consider this a viable workaround. Would still be great to fix the default value there and have some clarity on what the `libWwiseGDNative` framework's actual target OS version should be.

**Comment by alessandrofama** [comment]
Thanks for the report and for the detailed info. I don't have too much experience with iOS, but let's see if we can solve the problem. A few notes:

1) The Godot C++ bindings set the minimum deployment version for the static library we link to here in this build script to 10.0 (hardcoded): 
https://github.com/godotengine/godot-cpp/blob/f4f6fac4c784da8c973ade0dbc64a9d8400ee247/SConstruct#L222-L223

I don't know why it's not, but I think this value should be customizable by the user. My only guess is that since they are also targeting `armv7`, `10.0` would be the maximum deployment target for `armv7` (that's what Clang tells me!) and they wanted to have maximum compatibility in general.

If you change this to `11.0`, build and check with otool, the correct minimum version will be displayed:
```
% otool -l libgodot-cpp.ios.debug.arm64.a | grep -A4 VERSION

      cmd LC_VERSION_MIN_IPHONEOS
  cmdsize 16
  version 11.0
      sdk 15.4
Load command 2

```
2) For libWwiseGDNative.dylib we also set the deployment version to 10.0 here:
https://github.com/alessandrofama/wwise-godot-integration/blob/c6b4dfb5fc8437a62be932ffe5a32d89d9a97988/wwise-gdnative/SConstruct#L247-L250

We simply copied the relevant iOS part from the build script of godot-cpp here. Strangely, checking with otool does not show the command `LC_VERSION_MIN_IPHONEOS`, but instead the command `LC_BUILD_VERSION`:
```
% otool -l libWwiseGDNative.dylib | grep -A4 VERSION

      cmd LC_BUILD_VERSION
  cmdsize 32
 platform 2
    minos 15.4
      sdk 15.4
```
See `minos 15.4`, which matches your observation ("15.4 is the absolute last target OS version you could set at the time of writing"). It seems that I forgot to set the minimum version in the link flags:
```
env.Append(LINKFLAGS=['-miphoneos-version-min=11.0'])
```
Doing this will correctly set the minimum version:

```
% otool -l libWwiseGDNative.dylib | grep -A4 VERSION

      cmd LC_VERSION_MIN_IPHONEOS
  cmdsize 16
  version 11.0
      sdk 15.4
```

3) `libWwiseGDNative.framework` is generated by Godot when your export your project. `Info.plist` is generated here: https://github.com/godotengine/godot/blob/c02325ffda2258172385e70e160cd476fb19a004/platform/iphone/export/export.cpp#L1352-L1353

- This value is hardcoded and can only be changed manually as you have done. 

4) About editing `addons/wwise/bin/wwise-gdnative.gdnlib` for exporting the release libs. This should already work if you uncheck `Export with Debug` when exporting the project. Our [build export addon](https://github.com/alessandrofama/wwise-godot-integration/blob/main/wwise-gdnative/gdnative-demo/addons/wwise/editor/wwise_build_export.gd) should take care of this. If this doesn`t work, there is another bug I need to fix :)

So how to proceed?

- We can add a fix that allows users to choose the minimum deployment version for `libWwiseGDNative.dylib` when building the library. In your case (`11.0`), this would only work for the `arm64` target and would require a manual build of the library.
- We cannot change the Godot C++ bindings from hardcoding the `10.0` value. The workaround for this is manually editing the value in their build script (`SConstruct`) and to build the library yourself. We can try to report this as an issue in their repo.
- We cannot fix Godot hardcoding the `10.0` value in `Info.plist`. Again, the workaround here is to manually changing the value after exporting. Writing an issue in the godotengine repo might be possible for a future fix.
- I will provide updated arm64 libs below for you, targeting version `11.0`. If you have time, please try to replace yours, export only for arm64, change the `MinimumOSVersion` value to `11.0` and check if it works.
[iOS.zip](https://github.com/alessandrofama/wwise-godot-integration/files/8856174/iOS.zip)

**Comment by minosvasilias** [comment]
Hey Alessandro, apologies for the delay in replying, i must've missed the e-mail notifying me of this. And thanks a lot for the super detailed reply, this all makes sense it's extremely helpful to understand what exactly is happening. Appreciate the debug/release bug being fixed as well!

I was planning to do a bunch of builds tomorrow, so i will ty out the updated arm64 libs and let you know how it went. Thanks for getting these ready, i really appreciate that. Will respond here again once i have tested things.

**Comment by minosvasilias** [comment]
Hey, happy to report that the updated library you provided did indeed work perfectly! No error on app submission, and no other issues either. Thanks again for that, very helpful.

I fully agree with your observations otherwise, would make sense to have the minimum deployment target not hard-coded, but customizable. Providing that option when building this library would certainly be good, and i suppose updating the value used for future releases would make sense as well. 

Appreciate the work on this btw, this is has come in extremely useful for us and has been working a treat otherwise!  👍

**Comment by alessandrofama** [comment]
This was solved in Godot 4.0, as the new bindings allow a user-defined minimum iOS version to be specified. In Godot 3.5 it is still hard coded to 10.0 as it is today.
I will include the ability to build with a custom value when we update this integration for 4.0 and set the default value to 11.0.
